src/main/java/org/manifold/compiler/front/NodeValueVertex.java

```diff
@@ -0,0 +1,226 @@
+        futurePort = unwrapPort(inputPortTuple.entry("0"));
+        TupleValue attributesValue = (TupleValue) inputPortTuple.entry("1");
+        inputPortAttrs = attributesValue.getEntries();
+      }
+      futureInputPorts.put(inputPortName, futurePort);
+      portAttrs.put(inputPortName, inputPortAttrs);
+    }
+
+    // we must also get output port attributes, which appear as function inputs
+    for (String outputPortName : outputPortNames) {
+      PortTypeValue outputPortType = nodeType.getPorts().get(outputPortName);
+      Map<String, Value> outputPortAttrs;
+      if (outputPortType.getAttributes().isEmpty()) {
+        outputPortAttrs = new HashMap<>(); // no attributes
+      } else {
+        TupleValue attributesValue = (TupleValue) input.entry(outputPortName);
+        outputPortAttrs = attributesValue.getEntries();
+      }
+      portAttrs.put(outputPortName, outputPortAttrs);
+    }
+
+    for (String attrName : nodeType.getAttributes().keySet()) {
+      Value attrValue = input.entry(attrName);
+      nodeAttrs.put(attrName, attrValue);
+    }
+
+    node = new NodeValue(nodeType, nodeAttrs, portAttrs);
+
+  }
+
+  // Connect all input ports to their sources and return all connections formed.
+  public List<ConnectionValue> connect()
+      throws UndeclaredIdentifierException, UndeclaredAttributeException,
+      InvalidAttributeException, TypeMismatchException {
+    List<ConnectionValue> connections = new LinkedList<>();
+    for (Map.Entry<String, FuturePortValue> e : futureInputPorts.entrySet()) {
+      String inputPortName = e.getKey();
+      FuturePortValue source = e.getValue();
+      PortValue sourcePort = source.getPort();
+      PortValue targetPort = node.getPort(inputPortName);
+      // TODO connection attributes; for now we assume none
+      Map<String, Value> attrs = new HashMap<>();
+      ConnectionValue conn = new ConnectionValue(sourcePort, targetPort, attrs);
+      connections.add(conn);
+    }
+    return connections;
+  }
+
+  @Override
+  public void verify() throws Exception {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public boolean isElaborationtimeKnowable() {
+    // TODO Auto-generated method stub
+    return false;
+  }
+
+  @Override
+  public boolean isRuntimeKnowable() {
+    // TODO Auto-generated method stub
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    if (node == null) {
+      return "node (not elaborated)";
+    } else {
+      return "node (" + node.toString() + ")";
+    }
+  }
+
+  @Override
+  public void writeToDOTFile(BufferedWriter writer) throws IOException {
+    String objectID = Integer.toString(System.identityHashCode(this));
+    String label = this.toString();
+    writer.write(objectID);
+    writer.write(" [");
+    writer.write("label=\"");
+    writer.write(objectID);
+    writer.write("\n");
+    writer.write(label);
+    writer.write("\"");
+    writer.write("];");
+    writer.newLine();
+  }
+
+}
```

Comment:
I think a lot of this logic would disappear with #48. It seems like a lot of code to represent a single constant value...